### PR TITLE
[SYNPY-1175] Add functions to download the list of files from your download list

### PR DESCRIPTION
**Pull request by thomasyu888**
`get_download_list` downloads the manifest file, then loops through it to download all the files into the download list.  Everytime a file is successfully downloaded, it will remove it from the download list. Here is the [design doc](https://sagebionetworks.jira.com/wiki/spaces/SYNPY/pages/2547482760/Download+list+V2+interface) 

Here's how you test this branch:

1. Go on Synapse and add to your download list
2. Create a new Python virtualenv
3. Install this specific branch
    ```
    pip install pandas
    pip install https://github.com/thomasyu888/synapsePythonClient/archive/refs/heads/synpy-1175-download-list.zip
    ```
4. Test feature.
    ```
    import synapseclient
    syn = synapseclient.login()
    
    dl_list_file_entities = syn.get_download_list()
    ```
5. confirm that your download list on Synapse is empty

- [x] Add unit tests for low level API calls
- [x] Add tests for `get_download_list_manifest`
- [x] Add tests for `get_download_list`

### CODE REVIEW

File: synapseclient/client.py [hunk 1]
```diff
@@ -1354,6 +1355,103 @@ def uploadFileHandle(self, path, parent, synapseStore=True, mimetype=None, md5=N
...
+        # Download the manifest
+        downloaded_path = self._download_from_URL(
+            url=file_result['preSignedURL'],
+            destination="./",
+            fileHandleId=file_result['fileHandleId'],
+            expected_md5=file_result['fileHandle'].get('contentMd5')
+        )
+        return downloaded_path
+
+    def get_download_list(self) -> typing.List:
```
Comment (thomasyu888) [inline synapseclient/client.py:1426]: Could this function be better named?

**Comment by BrunoGrandePhD** [comment]
I don't have the bandwidth to review this one. Sorry!

File: synapseclient/client.py [hunk 2]
```diff
@@ -1354,6 +1355,115 @@ def uploadFileHandle(self, path, parent, synapseStore=True, mimetype=None, md5=N
         """
         return upload_file_handle(self, parent, path, synapseStore, md5, file_size, mimetype)
 
+    ############################################################
+    #                  Download List                           #
+    ############################################################
```
Comment (mattfazza) [inline synapseclient/client.py:1360]: I think it's a good idea to keep the naming of these functions consistent with the end point.

File: synapseclient/client.py [hunk 3]
```diff
@@ -1354,6 +1355,115 @@ def uploadFileHandle(self, path, parent, synapseStore=True, mimetype=None, md5=N
...
+                {fileEntityId: synid, versionNumber: version}
+
+        :returns: Number of files removed from download list
+        """
+        request_body = {"batchToRemove": list_of_files}
+        num_files_removed = self.restPOST(
+            "/download/list/remove",
+            body=json.dumps(request_body)
+        )
+        return num_files_removed
```
Comment (mattfazza) [inline synapseclient/client.py:1378]: What if the call to restPOST fails?

File: synapseclient/client.py [hunk 4]
```diff
@@ -1354,6 +1355,115 @@ def uploadFileHandle(self, path, parent, synapseStore=True, mimetype=None, md5=N
...
+            "concreteType": "org.sagebionetworks.repo.model.download.DownloadListManifestRequest",
+            "csvTableDescriptor": {
+                "separator": separator,
+                "quoteCharacter": quoteCharacter,
+                "escapeCharacter": escapeCharacter,
+                "lineEnd": lineEnd,
+                "isFirstLineHeader": header
+            }
+        }
+        return self._waitForAsync(uri="/download/list/manifest/async", request=request_body)
```
Comment (mattfazza) [inline synapseclient/client.py:1409]: Same as above.  Catching errors before returning could be a good implementation here.  Network requests are unpredictable (even when the API is made by us as well).

File: synapseclient/client.py [hunk 5]
```diff
@@ -1354,6 +1355,115 @@ def uploadFileHandle(self, path, parent, synapseStore=True, mimetype=None, md5=N
...
+        )
+        # Download the manifest
+        downloaded_path = self._download_from_URL(
+            url=file_result['preSignedURL'],
+            destination="./",
+            fileHandleId=file_result['fileHandleId'],
+            expected_md5=file_result['fileHandle'].get('contentMd5')
+        )
+        return downloaded_path
+
```
Comment (mattfazza) [inline synapseclient/client.py:1431]: The name of these functions are all fine to me.

**Review by mattfazza — APPROVED** [review]
These are suggestions.  All and all: this is a solid PR.  I would like to see more error handling though!

File: synapseclient/client.py (hunk 3 above)
Comment (thomasyu888) [reply synapseclient/client.py:1378]: the restPOST will return the respective errors

File: synapseclient/client.py (hunk 4 above)
Comment (thomasyu888) [reply synapseclient/client.py:1409]: There is error handling done in the client.  the respective Synapse errors will be returned

**Comment by thomasyu888** [comment]
Thanks @avanlinden for reviewing.
- [ ] Add ability to set downloadLocation 
- [ ] Make sure to add `path` column to the manifest file.

File: synapseclient/client.py [hunk 6]
```diff
@@ -1354,6 +1355,130 @@ def uploadFileHandle(self, path, parent, synapseStore=True, mimetype=None, md5=N
...
+                    entity = self.get(row['ID'], downloadLocation=downloadLocation)
+                    downloaded_files.append(entity.path)
+                    row['path'] = entity.path
+                    writer.writerow(row)
+                except Exception:
+                    print("Unable to download file")
+
+                # TODO: should the file be removed from the download list if it fails download
+                # Must include version number because you can have multiple versions of a
+                # file in the download list
```
Comment (thomasyu888) [inline synapseclient/client.py:1465]: Don't remove from filelist if I can't download

File: synapseclient/client.py [hunk 7]
```diff
@@ -1354,6 +1355,130 @@ def uploadFileHandle(self, path, parent, synapseStore=True, mimetype=None, md5=N
...
+            writer.writeheader()
+
+            for row in reader:
+                # You can add things to the download list that you don't have access to
+                # So there must be a try catch here
+                try:
+                    entity = self.get(row['ID'], downloadLocation=downloadLocation)
+                    downloaded_files.append(entity.path)
+                    row['path'] = entity.path
+                    writer.writerow(row)
```
Comment (thomasyu888) [inline synapseclient/client.py:1459]: Create a list of successfully downloaded files then all remove from download list at the very end after all downloads will succeed

File: synapseclient/client.py [hunk 8]
```diff
@@ -1354,6 +1355,130 @@ def uploadFileHandle(self, path, parent, synapseStore=True, mimetype=None, md5=N
...
+        # if manifest:
+        #     downloaded_files.append(dl_list_path)
+        # else:
+        #     os.remove(dl_list_path)
+        # Always remove original manifest file
+        os.remove(dl_list_path)
+        # Don't want to clear all the download list because you can add things
+        # to the download list after initiating this command
+        # self.clear_download_list()
+        return "SYNAPSE_METADATA_MANIFEST.tsv"
```
Comment (thomasyu888) [inline synapseclient/client.py:1480]: Always return path to manifest file

File: synapseclient/client.py (hunk 7 above)
Comment (thomasyu888) [reply synapseclient/client.py:1459]: Unsuccessful downloads should not be in the manifest file

File: synapseclient/client.py [hunk 9]
```diff
@@ -1354,6 +1355,130 @@ def uploadFileHandle(self, path, parent, synapseStore=True, mimetype=None, md5=N
...
+
+        :param manifest: Whether to keep the manifest file for the download list. Defaults to True.
+        :param downloadLocation: Directory to download files to.
+
+        :returns: manifest file with file paths
+        """
+        dl_list_path = self.get_download_list_manifest()
+        downloaded_files = []
+        with open(dl_list_path) as manifest_f, \
+             open('SYNAPSE_METADATA_MANIFEST.tsv', 'w') as write_obj:
```
Comment (thomasyu888) [inline synapseclient/client.py:1443]: Should be a csv file

Comment (thomasyu888) [reply synapseclient/client.py:1443]: Should be manifest.csv

**Pull request merged**